### PR TITLE
Implement Compose

### LINKED ISSUE #23: Implement Compose
Like in audiomentations: https://github.com/iver56/audiomentations/blob/master/audiomentations/core/composition.py#L4

### PR DESCRIPTION
**Pull request by iver56**
Close #23

### CODE REVIEW

File: torch_audiomentations/core/composition.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    ):
+        """
+
+        :param transforms: List of waveform transform instances
+        :param shuffle: Should the order of transforms be shuffled?
+        :param p: The probability of applying the Compose to the given batch.
+        :param p_mode: Only "per_batch" is supported at the moment.
+        """
+        super().__init__()
+        self.transforms = transforms
```
Comment (hbredin) [inline torch_audiomentations/core/composition.py:25]:
```suggestion
        self.transforms = torch.nn.ModuleList(transforms)
```
otherwise we'd loose differentiability, wouldn't we?
Maybe add a test in https://github.com/asteroid-team/torch-audiomentations/blob/master/tests/test_differentiable.py?

File: torch_audiomentations/core/composition.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        self.p = p
+        if p_mode != "per_batch":
+            # TODO: Support per_example as well? And per_channel?
```
Comment (hbredin) [inline torch_audiomentations/core/composition.py:28]: I am not sure what would happen when `Compose` is `per_something` and sub-transforms are `per_something_else`.

File: torch_audiomentations/core/composition.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            raise Exception('p_mode = "{}" is not supported')
+        self.p_mode = p_mode
+        self.shuffle = shuffle
+
+        name_list = []
+        for transform in self.transforms:
+            name_list.append(type(transform).__name__)
+        self.__name__ = "_".join(name_list)
```
Comment (hbredin) [inline torch_audiomentations/core/composition.py:36]: Should `__name__` somehow reflect the state of `shuffle`?
Maybe `__name__` is not that important anyway...

For instance, though using emojis is probably not a good idea:
* `Gain ⏩ PolarityInversion` when `shuffle` is `False` 
* `Gain 🔀 PolarityInversion` when `shuffle` is `True`

File: tests/test_compose.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
+import unittest
+
+import numpy as np
+import torch
+from numpy.testing import assert_almost_equal, assert_array_equal
+
+from torch_audiomentations import PolarityInversion, Compose, PeakNormalization
+from torch_audiomentations.augmentations.gain import Gain
```
Comment (hbredin) [inline tests/test_compose.py:8]: Why isn't it imported from `torch_audiomentations` like the other ones?

File: tests/test_compose.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+from torch_audiomentations.utils.dsp import convert_decibels_to_amplitude_ratio
+
+
+class TestCompose(unittest.TestCase):
+    def test_compose(self):
+        samples = np.array([[[1.0, 0.5, -0.25, -0.125, 0.0]]], dtype=np.float32)
+        sample_rate = 16000
+
+        augment = Compose(
+            [
```
Comment (hbredin) [inline tests/test_compose.py:18]: Nitpicking...
```suggestion
            transforms=[
```

File: tests/test_compose.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+            # Either PeakNormalization or Gain was applied last
+            if processed_samples[0, 0, 0] < 0.2:
+                num_gain_last += 1
+            elif processed_samples[0, 0, 0] == 1.0:
+                num_peak_normalization_last += 1
+            else:
+                raise AssertionError("Unexpected value!")
+
+        self.assertGreater(num_peak_normalization_last, 20)
```
Comment (hbredin) [inline tests/test_compose.py:105]: Risky but this will hopefully be OK once 2020 is behind us.

File: README.md [hunk 7]
```diff
@@ -18,24 +18,30 @@ Audio data augmentation in PyTorch. Inspired by [audiomentations](https://github
...
+from torch_audiomentations import Compose, Gain, PolarityInversion
 
 
 # Initialize augmentation callable
-apply_gain_augmentation = Gain(
-    min_gain_in_db=-15.0,
-    max_gain_in_db=5.0,
-    p=0.5,
+apply_augmentation = Compose(
+    [
```
Comment (hbredin) [inline README.md:26]: Nitpicking...
```suggestion
    transforms=[
```

File: scripts/demo.py [hunk 8]
```diff
@@ -78,6 +78,32 @@ def __exit__(self, type, value, traceback):
     modes = ["per_batch", "per_example", "per_channel"]
     for mode in modes:
         transforms = [
+            {
+                "instance": Compose(
+                    [
```
Comment (hbredin) [inline scripts/demo.py:83]: Nitpicking...
```suggestion
                    transforms=[
```

File: scripts/demo.py [hunk 9]
```diff
@@ -78,6 +78,32 @@ def __exit__(self, type, value, traceback):
...
+                        PeakNormalization(mode=mode, p=1.0),
+                    ],
+                    shuffle=True,
+                ),
+                "name": "Shuffled Compose with Gain and PeakNormalization",
+                "num_runs": 5,
+            },
+            {
+                "instance": Compose(
+                    [
```
Comment (hbredin) [inline scripts/demo.py:96]: Nitpicking...
```suggestion
                    transforms=[
```

File: torch_audiomentations/core/composition.py (hunk 2 above)
Comment (iver56) [reply torch_audiomentations/core/composition.py:28]: Let's say you have 8 audio snippets in a batch, and you want to apply the Compose to 50% of them. Then that would be `p=0.5, p_mode="per_example"` (which is not implemented yet). Each transform has its own mode and p_mode that is used independently from other transforms

Comment (iver56) [reply torch_audiomentations/core/composition.py:28]: I agree that it doesn't make sense to add `mode=` in Compose

File: torch_audiomentations/core/composition.py (hunk 3 above)
Comment (iver56) [reply torch_audiomentations/core/composition.py:36]: Yeah, I don't know if this is important. It's copy-pasta from audiomentations 🙈 Should I remove it?

File: tests/test_compose.py (hunk 4 above)
Comment (iver56) [reply tests/test_compose.py:8]: Oh, that's probably an error. Let me fix it.

File: torch_audiomentations/core/composition.py (hunk 3 above)
Comment (hbredin) [reply torch_audiomentations/core/composition.py:36]: Depends on whether you had any use for this `__name__` thing in mind?

File: tests/test_compose.py (hunk 6 above)
Comment (iver56) [reply tests/test_compose.py:105]: Hmm, according to math, this could indeed fail once every ~7 billion times. Should I set a fixed random seed?

https://www.wolframalpha.com/widgets/view.jsp?id=78baf4f3a070cc5b9b226664d2ce80ec

File: tests/test_compose.py (hunk 5 above)
Comment (iver56) [reply tests/test_compose.py:17]: Some people will use an arg, some people will use a kwarg :) I added `transforms=` in some of the tests now, so we cover both

File: torch_audiomentations/core/composition.py (hunk 3 above)
Comment (iver56) [reply torch_audiomentations/core/composition.py:36]: Nope. I think I'll remove it for now

File: torch_audiomentations/core/composition.py (hunk 1 above)
Comment (iver56) [reply torch_audiomentations/core/composition.py]: Thanks, nice catch! Done.

File: torch_audiomentations/core/composition.py (hunk 2 above)
Comment (iver56) [reply torch_audiomentations/core/composition.py:28]: I left it as a TODO for later: https://github.com/asteroid-team/torch-audiomentations/issues/34

**Review by hbredin — APPROVED** [review]

File: torch_audiomentations/core/composition.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        :param shuffle: Should the order of transforms be shuffled?
+        :param p: The probability of applying the Compose to the given batch.
+        :param p_mode: Only "per_batch" is supported at the moment.
+        """
+        super().__init__()
+        self.transforms = torch.nn.ModuleList(transforms)
+        self.p = p
+        if p_mode != "per_batch":
+            # TODO: Support per_example as well? And per_channel?
+            raise Exception('p_mode = "{}" is not supported')
```
Comment (mpariente) [inline torch_audiomentations/core/composition.py:29]:
```suggestion
            raise ValueError(f'p_mode = "{p_mode}" is not supported')
```

Comment (iver56) [reply torch_audiomentations/core/composition.py:29]: Thanks
